### ISSUE TITLE
Group Location Issue

### ISSUE THREAD

**Issue by symbioquine** [symbioquine/farmOS_asset_link#44]
>  ![image](https://github.com/symbioquine/farmOS_asset_link/assets/30754460/d08b291a-1bab-4be6-9c5d-a8248ff2def5)
> I did confirm an issue with group memberships and asset link. It appears that asset link is not taking into account the assets location if it's in a group.and the group has a location.
> 
> These are screenshots of the 2 animal assets and the group asset they belong to in farmOS
>
> ![image](https://github.com/symbioquine/farmOS_asset_link/assets/30754460/2db17897-67bf-4ac5-a0aa-6076b70d49e8)
> ![image](https://github.com/symbioquine/farmOS_asset_link/assets/30754460/40e55dce-98c5-492f-b076-62ac8e676b7f)
> ![image](https://github.com/symbioquine/farmOS_asset_link/assets/30754460/f65441e8-08f4-4492-b540-f5254311d4c0)
> 
> Here is the same assets in asset link.
>
> ![image](https://github.com/symbioquine/farmOS_asset_link/assets/30754460/65153e0b-d425-4bb6-a92d-ddfc20e34783)
> ![image](https://github.com/symbioquine/farmOS_asset_link/assets/30754460/a73e91e3-4c5e-481c-ac02-46f388d55c4c)
> ![image](https://github.com/symbioquine/farmOS_asset_link/assets/30754460/d9a2d953-c515-481d-80d4-cd694fc08cbb)

**Comment by mstenta** [comment]
Is asset link computing location itself?

The Group module overrides the asset location logic to take into account group membership.

Notably, assets have a computed `location` and `geometry` in the farmOS API. Perhaps Asset Link should load that instead of computing location itself. Then it will always have the same logic.

Although I imagine the reason for computing internally is so that it can work offline? So if you move an animal while you're offline it will show the correct location in Asset Link... is that the intention?

**Comment by mstenta** [comment]
> So if you move an animal while you're offline it will show the correct location in Asset Link... is that the intention?

One idea: maybe if Asset Link detects that you are offline, it puts a little warning symbol next to computed fields like Location to indicate that they may not be updated until network connection is re-established.

(This assumes that it moves to using the computed `location` field of assets from the API rather than computing it itself.)

**Comment by mstenta** [comment]
> The Group module overrides the asset location logic to take into account group membership.

Of course, Asset Link *could* also add logic to mimic the Group module's... but that might be unreasonable to maintain. And it's always possible that other modules could change it in other ways. :sweat_smile:

**Comment by symbioquine** [comment]
> Although I imagine the reason for computing internally is so that it can work offline? So if you move an animal while you're offline it will show the correct location in Asset Link... is that the intention?

Yes that's the goal.

Though Asset Link tries not to re-compute the group/location/geometry information unless it has reason to "believe" it needs to: https://github.com/symbioquine/farmOS_asset_link/blob/a6b8d2b79e965abd4ad7846e524535c85f2864cd/alinkjs/packages/assetlink/src/FarmDataModelOrbitMemorySourceDecorator.js#L321-L326

The problems are two-fold;

1. I can't remember whether I've implemented that transitive group location/movement logic
2. Even if I have, this could easily be a cache invalidation problem - which are more-or-less inevitable (especially as the depth of relationships that have to be queried grows

**Comment by symbioquine** [comment]
> One idea: maybe if Asset Link detects that you are offline, it puts a little warning symbol next to computed fields like Location to indicate that they may not be updated until network connection is re-established.

This is actually one area I'm really looking forward to improving Asset Link in...

I would like to provide a more powerful API for interacting with the cache status/info of individual parts of a given farmOS api response. Further I'd also like to make it easier to implement UIs that show the cached data immediately, but (when online) also request the latest data from the server and then surface any new data gracefully (without jumping around as it loads).

**Comment by symbioquine** [comment]
> 1. I can't remember whether I've implemented that transitive group location/movement logic

From a quick skim of the code, I think the answer is that I haven't.
